src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,960 @@
+        // TODO: improve this
+        next_node.cost = node.cost + 10 + ( steer ? 1 : 0 );
+        next_node.speed = next_speed;
+        next_node.target_speed_tps = target_speed_tps;
+        next_nodes.emplace_back( next_addr, next_node );
+    }
+}
+
+cata::optional<std::vector<navigation_step>> vehicle::autodrive_controller::compute_path(
+            int speed_tps ) const
+{
+    if( speed_tps == 0 || speed_tps < -1 ) {
+        return cata::nullopt;
+    }
+    // TODO: tweak this
+    constexpr int max_search_count = 10000;
+    std::vector<navigation_step> ret;
+    // TODO: check simple reachability first and bail out or set upper bound on node score
+    std::unordered_map<node_address, navigation_node, node_address_hasher> known_nodes;
+    std::priority_queue<scored_address, std::vector<scored_address>, std::greater<>>
+            open_set;
+    const tripoint_abs_ms nav_corner_pos = project_to<coords::ms>( data.current_omt ) - point(
+            NAV_MAP_PADDING, NAV_MAP_PADDING );
+    const point start_pos = ( abs_vehicle_pos( driven_veh ) - nav_corner_pos ).raw().xy();
+    const node_address start = make_node_address( start_pos, driven_veh.face.dir() );
+    known_nodes.emplace( start, make_start_node( start, driven_veh ) );
+    open_set.push( scored_address{ start, 0 } );
+    int search_count = 0;
+    std::vector<std::pair<node_address, navigation_node>> next_nodes;
+    while( !open_set.empty() ) {
+        const node_address cur_addr = open_set.top().addr;
+        open_set.pop();
+        search_count++;
+        if( data.goal_zone.contains( point{ cur_addr.x, cur_addr.y } ) ) {
+            node_address addr = cur_addr;
+            while( !( addr == start ) ) {
+                const navigation_node &node = known_nodes.at( addr );
+                const node_address &prev = node.prev;
+                ret.emplace_back( navigation_step{
+                    nav_corner_pos + point( prev.x, prev.y ),
+                    addr.facing_dir,
+                    node.target_speed_tps *VMIPH_PER_TPS
+                } );
+                addr = prev;
+            }
+            return ret;
+        }
+        const navigation_node &cur_node = known_nodes[cur_addr];
+        next_nodes.clear();
+        compute_next_nodes( cur_addr, cur_node, speed_tps, next_nodes );
+        for( const auto &next : next_nodes ) {
+            const node_address &next_addr = next.first;
+            const navigation_node &next_node = next.second;
+            auto iter = known_nodes.find( next_addr );
+            if( iter != known_nodes.end() ) {
+                const navigation_node &other_node = iter->second;
+                if( next_node.cost < other_node.cost ) {
+                    const bool same_dir = cur_addr.facing_dir == next_addr.facing_dir;
+                    if( ( !same_dir && next_node.tileray_steps == 1 )
+                        || next_addr.facing_dir % 3 == 0 ) {
```

Comment:
minor stylistic nit: binary operators go on the proceeding line, so
```c++
                    if( ( !same_dir && next_node.tileray_steps == 1 ) ||
                         next_addr.facing_dir % 3 == 0 ) {
```

Reply:
I just let astyle handle the formatting. The code's changed a bit so it doesn't look like this in my working copy (which I need to push soon but I'm trying to find an off-by-one error).